Partial fix for restore snapshot GET /vm/config bug

summerlinryan opened the pull request:
Fixed the issue where GET /vm/config would return the default config for a VM restored from a snapshot. The config now updates after restoring from a snapshot.

# Reason for This PR

This PR exists to fix the bug described in issue #2783. 

## Description of Changes

Changed restoring a snapshot to include updating the `VmResources` so that the correct config would be returned for the GET /vm/config endpoint. The only parts of the config that are still using default values after restoring a snapshot are the following:  boot-source (which is reasonable since there is no boot-source in the case of a restore), machine-config.smt and machine-config.cpu_template. 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

summerlinryan commented:
@acatangiu @alindima I've left TODOs that describe the parts that I'm still confused on as of now. Sorry there are more TODOs than I was hoping for, just trying to learn and make more sense of how Firecracker all works together.

alindima commented:
@summerlinryan sorry, I haven't found the time to look over the PR yet, but will do it this week! Thanks!

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -153,6 +159,67 @@ impl VmResources {
+                .map_err(Error::BalloonDevice)?;
+        }
+
+        vmm.mmio_device_manager.for_each_virtio_device(|virtio_type, _device_id, _device_info, virtio_device| {
```
We can limit the calls to `for_each_virtio_device` to just one, by `match`-ing on the `virtio_type` inside the closure

summerlinryan replied:
That makes perfect sense. I'll definitely do that :+1:

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -153,6 +159,67 @@ impl VmResources {
 
+    /// Configures VmResources from the state of the Vmm.
+    pub fn from_vmm(vmm: &Vmm) -> std::result::Result<Self, Error> {
+        let mut resources = Self::default();
+        resources.boot_config = None;
+
+        if let Ok(balloon_config) = vmm.balloon_config() {
+            resources
+                .set_balloon_device(BalloonDeviceConfig::from(balloon_config))
+                .map_err(Error::BalloonDevice)?;
+        }
+
+        vmm.mmio_device_manager.for_each_virtio_device(|virtio_type, _device_id, _device_info, virtio_device| {
+            if virtio_type != TYPE_BLOCK {
+                return Ok(());
+            }
+
+            resources.set_block_device(BlockDeviceConfig::from(
+                virtio_device
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_any()
+                    .downcast_ref::<Block>()
+                    .unwrap()
+            ))
+            .map_err(Error::BlockDevice)
+        })?;
```
The methods here are actually reconstructing the devices. We only need to push the existing ones to the `Builder` objects. You will need to add some helper methods on the Builder objects for this

summerlinryan replied:
So this piece of feedback is the one that I'm having the hardest time implementing. For the `BlockBuilder` I was thinking about adding a new method like this:

```
pub fn insert_block(&mut self, block: &Block) {
    if block.is_root_device() {
        self.list.push_front(Arc::new(Mutex::new(block.clone())));
    } else {
        self.list.push_back(Arc::new(Mutex::new(block.clone())));
    }
}
```
which would allow using something like the following when setting `VmResources` from a snapshot

```
self.set_block_device(
    virtio_device
        .lock()
        .expect("Poisoned lock")
        .as_any()
        .downcast_ref::<Block>()
        .unwrap()
        .to_owned()
);
```

where `self` is the `VmResources` and the `virtio_device` is one of the arguments of the closure in the `for_each_virtio_device`. The `set_block_device` above would be a method that just calls the `insert_block` on the `BlockBuilder` (the existing `set_block_device` on `VmResources` would be renamed to `set_block_device_from_config` since it uses a `BlockDeviceConfig`). However, this doesn't work because it'd either require that Block satisfies the Clone trait would require a ton of other types inside of Block satisfying the Clone trait which I was thinking wouldn't be the correct approach.

acatangiu replied:
What you need to do get a hold of the `Arc<>` part of `Arc<Mutex<Device>>` so that you can clone the reference without also cloning the device.

The problem is that the `MMIODeviceManager::restore` builds devices internally and you never see the `Arc`. Furthermore the `MMIODeviceManager` interface doesn't offer any API to expose the `Arc`: there is only https://github.com/firecracker-microvm/firecracker/blob/83ec483d4254936e1e2fd4adfe94761611a74a93/src/devices/src/bus.rs#L98 which provides only plain reference

So I see two options:
1. Change `MMIODeviceManager::get_device()` and `Bus::get_device()` APIs to return `Option<(u64, Arc<Mutex<dyn BusDevice>>)>`, the you can change the passed function signature in `MMIODeviceManager::for_each_device()` to also take `Arc<Mutex<dyn BusDevice>>` so that within the closure you can clone the `Arc` reference to the device and thus share ownership of it.
2. When restoring devices from states using the `MMIODeviceManager::restore` here https://github.com/firecracker-microvm/firecracker/blob/83ec483d4254936e1e2fd4adfe94761611a74a93/src/vmm/src/builder.rs#L528 you can also provide a function/closure parameter to `MMIODeviceManager::restore` to do auxiliary work on the restored devices (like in your case, also adding them to VmResources). The new signature could look like:
    ```rust
    fn restore(
        constructor_args: Self::ConstructorArgs,
        state: &Self::State,
        for_each_device: Fn(Arc<Mutex<dyn BusDevice>>),
    ) -> Result<Self, Self::Error> {
    ```

I think first option has less cognitive complexity.

summerlinryan replied:
Sorry again for being slow to respond I got busy over the holidays.

TL;DR: How do I cast a `Arc<Mutex<dyn VirtioDevice>>` into a `Arc<Mutex<Block>>`?

I'm starting to wonder if I just don't have a deep enough understanding of Rust to work on a project as sophisticated as Firecracker. If helping guide me through this issue becomes too burdensome, just let me know if it's best if someone else takes over this bug and I'll completely understand (I can try to contribute again in the future to a simpler problem or when I have a better understanding of Rust/Firecracker).

If you guys are able to continue helping, I'm at a point where I have modified the `MMIODeviceManager::get_device()` and `Bus::get_device()` APIs to return the actual `Arc<Mutex<dyn BusDevice>`, I've modified the closure of `MMIODeviceManager::for_each_virtio_device` to take a `Arc<Mutex<dyn VirtioDevice>>` and I've modified the closure of  `MMIODeviceManager::for_each_device` to take a `Arc<Mutex<dyn BusDevice>>`. The problem I'm having now is that I have a method that looks like 

```rust
pub fn set_block_device(&mut self, block_device: Arc<Mutex<Block>>>)
```
on the `struct VmResources` but the when I clone the `Arc<Mutex<dyn VirtioDevice>>` inside of the `MMIODeviceManager::for_each_device` closure, it can't be passed to `set_block_device` as a `Arc<Mutex<Block>>` since the types don't match. I've tried downcasting the `dyn VirtioDevice` into the `Block` inside of the `Arc<Mutex<dyn VirtioDevice>>` but that seems to cast the entire expression to a `&Block` rather than a `Arc<Mutex<Block>>` which is what I need.

acatangiu replied:
Hi @summerlinryan,

> I'm starting to wonder if I just don't have a deep enough understanding of Rust to work on a project as sophisticated as Firecracker.

Don't beat yourself up, this issue is not a straightforward issue. But it is a good learning opportunity! :+1: 

You are right, I don't know if it's possible to clone `Arc<Mutex<dyn Trait>>` into `Arc<Mutex<T: Trait>>`.

The other way around is easily allowed through type coercion (`let t1 = Arc::clone(&struct) as Arc<Mutex<Trait>>;`) because correctness can be easily checked at compile time.

I believe it could be done by implementing `Any` trait for `Mutex<dyn BusDevice>`, `Mutex<Block>`, etc. Then you can use [Arc::downcast](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.downcast) to get what you want.

Having said that, you could also do the simpler thing and go with option 2 from [here](https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r774466266) with a simple `enum` that gets rid of all dynamic casting:
```rust
enum SharedDeviceTypeWrapper {
    SharedBlock(Arc<Mutex<Block>>),
    SharedNetwork(Arc<Mutex<Net>>),
    SharedBalloon(Arc<Mutex<Balloon>>),
    SharedEtc(Arc<Mutex<Etc>>),
}

fn MMIODeviceManager::restore(
    constructor_args: Self::ConstructorArgs,
    state: &Self::State,
    for_each_restored_device: FnOnce(enum SharedDeviceTypeWrapper),
) -> Result<Self, Self::Error>
{
    // ...
    if let Some(balloon_state) = &state.balloon_device {
        let device = Arc::new(Mutex::new(
            Balloon::restore(
                BalloonConstructorArgs { mem: mem.clone() },
                &balloon_state.device_state,
            )
            .map_err(Error::Balloon)?,
        ));

        // CALL NEW FN HERE:
        for_each_restored_device(SharedDeviceTypeWrapper::SharedBalloon(device.clone()));
        // same code as before after this ^

        restore_helper(
            device.clone(),
            device,
            &balloon_state.device_id,
            &balloon_state.transport_state,
            &balloon_state.mmio_slot,
            constructor_args.event_manager,
        )?;
    }
    // ...
}
```

`pub fn build_microvm_from_snapshot(` could take a `vm_resources: &super::resources::VmResources` reference, and in `build_microvm_from_snapshot` have smth like:
```rust
    // ...
    let register_to_resources = |shared_dev: enum SharedDeviceTypeWrapper| {
        match shared_dev {
            SharedDeviceTypeWrapper::SharedBalloon(b) => vm_resources.register_existing_balloon_device(b),
            ...
        };
    };
    // ...
    vmm.mmio_device_manager =
        MMIODeviceManager::restore(mmio_ctor_args, &microvm_state.device_states, register_to_resources)
            .map_err(MicrovmStateError::RestoreDevices)
            .map_err(RestoreMicrovmState)?;
```

`register_to_resources()` closure in example above could even be a member function of `VmResources` instead of a closure defined in `builder.rs`.

summerlinryan replied:
I started adding the `for_each_restored_device` function to `MMIODeviceManager::restore` but I noticed that requires adding it to the `Persist` trait which is implemented by a lot of different structs across many different crates. To do so would require either adding a ton of different crate dependencies to allow all of those different crates to know about the new `SharedDeviceTypeWrapper` enum. It would also require adding the new closure argument to all places that currently call the `Persist::restore` method which is quite a few and most of them don't really need to specify the closure anyway since it's primarily just used for setting resources when restoring from a snapshot in `build_microvm_from_snapshot`

Unless I'm misunderstanding the above, then I'm thinking that maybe the next approach to try is to implement Any for `Mutex<dyn VirtioDevice>` like you mentioned, so I'll try that unless you have a workaround to the above problems and still prefer the `for_each_restored_device` approach.

acatangiu replied:
I think you can add it to `MMIODevManagerConstructorArgs` [here](https://github.com/firecracker-microvm/firecracker/blob/73ccdfd643fd60b7c22f06fd05765470de3102b2/src/vmm/src/builder.rs#L522) so that you don't have to change the trait (definitely don't want to change that generic trait with this specific param).

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -153,6 +159,67 @@ impl VmResources {
+            track_dirty_pages: vmm
+                .guest_memory()
+                .iter()
+                .any(|region| (region.flags() as u32) & KVM_MEM_LOG_DIRTY_PAGES != 0)
```
a nicer way of doing this is to: `.all(|region| (region.bitmap().is_some())`

alindima commented on src/vmm/src/rpc_interface.rs:
```diff
@@ -441,20 +441,23 @@ impl<'a> PrebootApiController<'a> {
-        .and_then(|vmm| {
+        ).and_then(|(vm_resources, vmm)| {
+            println!("{}", serde_json::to_string(&VmmConfig::from(&vm_resources)).unwrap());
+            // TODO: how to deal with lifetime issue here
```
The problem here is that you want to take a reference to an owned vmm variable, but the vmm variable is local to the closure. So the reference `self.vm_resources` would live longer than the actual object.

Instead of returning the `VmResources` from `restore_from_snapshot`, you should pass it in there as a mutable reference and modify that variable to contain the device information.

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -153,6 +159,67 @@ impl VmResources {
+            Ok(())
+        })?;
+
+        // TODO: how to determine vsock & mmds config (seems like should be able to use for_each_virtio_device
```
I also noticed that currently (even if not resumed from a snapshot), the outputted MMDS config is `None` if a custom IP address has not been specified but a net device was configured with `allow_mmds_requests: true`. We should actually output the default MMDS config in this case.
You could include this fix in the PR.

summerlinryan replied:
> the outputted MMDS config is None if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true. We should actually output the default MMDS config in this case.

I'm a little confused by this; could you elaborate? I see the mmds-config appears to be always be null when hitting the `GET /vm/config` endpoint which I guess is what you're saying is a problem. Specifically the part I'm confused on is what you mean by "if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true". I'm reading through the MMDS documentation again to see if I can make more sense of what you mean by that.

alindima replied:
> I'm a little confused by this; could you elaborate? I see the mmds-config appears to be always be null when hitting the GET /vm/config endpoint which I guess is what you're saying is a problem. Specifically the part I'm confused on is what you mean by "if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true". I'm reading through the MMDS documentation again to see if I can make more sense of what you mean by that.

If a user does not customise the default IP address used for MMDS, there is no MMDS IP address outputted when dumping the vm config (while there is actually a default one).

We could open a separate issue for this, you don't need to tackle it in this PR. Let's take them one step at a time. WDYT?

alindima replied:
This will be solved when implementing the proposal from this comment: https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r812532289

summerlinryan commented:
Thanks for the feedback! Just wanted to let you know that I've started working through implementing the suggestions and hope to have some more code pushed within the next day or two.

summerlinryan commented on resources/seccomp/x86_64-unknown-linux-musl.json:
```diff
@@ -46,6 +46,18 @@
                 "comment": "Used for drive patching & rescanning, for reading the local timezone from /etc/localtime"
             },
+            {
+                "syscall": "stat",
+                "comment": "Used for restoring from snapshot (setting vm resources of api controller)"
+            },
+            {
+                "syscall": "timerfd_create",
+                "comment": "Used for restoring from snapshot (setting vm resources of api controller)"
+            },
+            {
+                "syscall": "eventfd2",
+                "comment": "Used for restoring from snapshot (setting vm resources of api controller)"
```
I'm still trying to make full sense of what's happening with this but if I remove the above syscalls then I get the errors related to "intercepting signal 31, code 1 and a bad syscall(4)" when trying to load the snapshot with the PUT /snapshot/load route.

alindima replied:
hm I'm fairly sure they were caused by the fact that the code was reconstructing the devices when populating the `VmResources` struct. If we no longer do this we should be able to remove them

alindima replied:
let's now remove these as well

summerlinryan commented on src/vmm/src/resources.rs:
```diff
@@ -153,6 +159,67 @@ impl VmResources {
+        resources.set_vm_config(&VmConfig {
+            vcpu_count: Some(vmm.vcpus_handles.len() as u8),
+            mem_size_mib: Some(mem_size_mib(vmm.guest_memory()) as usize),
+            ht_enabled: Some(false), // TODO: how to determine if HT enabled?
+            cpu_template: None, // TODO: how to determine cpu_template?
```
Sounds good, I'll update the swagger endpoint description to indicate that those pieces of information will be inaccurate when using GET /vm/config with a restored VM. For the log message, are you referring to just using `println!` to log some information whenever the `GET /vm/config` endpoint is hit or are you referring to updating some documentation?

alindima replied:
similarly to a `println!`, use the `warn!` macro which is used for logging warnings.
(There are other macros in the logger crate as well, for `error!`, `info!`, `debug!`)

alindima commented:
> Thanks for the feedback! Just wanted to let you know that I've started working through implementing the suggestions and hope to have some more code pushed within the next day or two.

Great! Thank you for the effort! Looking forward to the updates

summerlinryan commented:
Well, finally some progress has been made from an implementation standpoint! :+1:  It looks like all of the devices are showing correctly in the config in `GET /vm/config` after restoring from a snapshot. I'm sure there is still plenty of room for feedback but it's good to at least see things working. The only thing that isn't working is the mmds config which I've left another comment for in response to what Alin was saying about the mmds config. 

@acatangiu I ended up using just a plain function pointer instead of the FnOnce trait for the `for_each_restored_device` function because I was running into some issues with FnOnce not being `Sized` and therefore the `MMIODevManagerConstructorArgs` wasn't a valid parameter for the `restore` method. Not sure if a plain function pointer is acceptable here or not. If not, I can revisit trying to get it to work with FnOnce, maybe by making the constructor args generic like `MMIODevManagerConstructorArgs<F: FnOnce(&mut self VmResources, SharedDeviceType)`?

The next thing I'm moving onto is adding some tests and making all existing tests pass. I'm thinking unit tests won't be enough and I'll probably need to integration tests for this since it involves a VM being restored from a snapshot.

alindima commented on src/api_server/swagger/firecracker.yaml:
```diff
@@ -604,7 +604,8 @@ paths:
       summary: Gets the full VM configuration.
       description:
-        Gets configuration for all VM resources.
+        Gets configuration for all VM resources. If the VM is restored from a snapshot the boot-source,
+        machine-config.ht_enabled, and machine-config.cpu_template will all have default values.
```
```suggestion
        Gets configuration for all VM resources. If the VM is restored from a snapshot, the boot-source,
        machine-config.ht_enabled and machine-config.cpu_template will be empty.
```

They should be empty, since they don't really have default values

summerlinryan replied:
I figured this comment meant that boot-source should be null when a snapshot is restored instead of a default full object value. Is that what you meant? I made boot-source null in this [commit](https://github.com/firecracker-microvm/firecracker/pull/2806/commits/3d652cc2d8af7b396e0140812172277976bb7ad6).

alindima commented on src/vmm/src/builder.rs:
```diff
@@ -556,7 +565,22 @@ pub fn build_microvm_from_snapshot(
     .map_err(StartMicrovmError::Internal)?;
 
-    Ok(vmm)
+    {
+        let vmm = vmm_reference.lock().expect("Poisoned lock");
+
+        vm_resources.set_vm_config(&VmConfig {
+            vcpu_count: Some(vmm.vcpus_handles.len() as u8),
+            mem_size_mib: Some(mem_size_mib(vmm.guest_memory()) as usize),
+            ht_enabled: None,
+            cpu_template: None,
+            track_dirty_pages: vmm
+                .guest_memory()
+                .iter()
+                .any(|region| (region.bitmap().is_some()))
+        }).map_err(SetVmResources)?;
+    }
```
I'd move this a bit higher up in the function body, and remove the extra code block. We can also use the `vmm` variable that is already in scope for this function

alindima commented on src/vmm/src/builder.rs:
```diff
@@ -556,7 +565,22 @@ pub fn build_microvm_from_snapshot(
+        let vmm = vmm_reference.lock().expect("Poisoned lock");
+
+        vm_resources.set_vm_config(&VmConfig {
+            vcpu_count: Some(vmm.vcpus_handles.len() as u8),
```
There is already a `vcpu_count` variable in scope that we may use. Same for `track_dirty_pages` and `guest_memory`

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
-#![deny(warnings)]
+// #![deny(warnings)]
```
let's remove this if there aren't any warnings to suppress

alindima commented on src/vmm/src/vmm_config/net.rs:
```diff
@@ -147,6 +147,21 @@ impl NetBuilder {
+    /// Adds an existing network device in the builder.
+    pub fn add_device(&mut self, device: Arc<Mutex<Net>>) {
+        if let Some(index) = self
+            .net_devices
+            .iter()
+            .position(|net| {
+                net.lock().expect("Poisoned lock").id() == &device.lock().expect("Poisoned lock").id
+            })
+        {
+            self.net_devices.swap_remove(index);
+        }
```
When restoring from a snapshot there will never be two net devices with the same id. The Firecracker VM on which the snapshot was taken would have prevented this at net device creation time.

alindima reviewed (commented):
Nice work!

alindima commented:
> Not sure if a plain function pointer is acceptable here or not. If not, I can revisit trying to get it to work with FnOnce, maybe by making the constructor args generic like MMIODevManagerConstructorArgs<F: FnOnce(&mut self VmResources, SharedDeviceType)?

I think this is perfectly fine. This is not a closure that captures variables from the outside scope so a plain function pointer works just right. @acatangiu agree?

> The next thing I'm moving onto is adding some tests and making all existing tests pass. I'm thinking unit tests won't be enough and I'll probably need to integration tests for this since it involves a VM being restored from a snapshot.

Yes, try to add unit tests that cover all the new code. Python integration tests will also be needed to ensure that everything works end-to-end.

Thanks for the hard work! This is no easy issue and I think we're on the right track.

summerlinryan commented:
A few questions: 

1. I noticed a coverage test passing and I feel like I recall reading somewhere in the documentation a while back that mentioned a need to update some coverage numbers in order to get some tests to pass. I can't seem to find any detailed information in the docs about updating the coverage numbers. Any idea there is some documentation describing what coverage numbers need to be updated and how?
2. For integration tests, I've seen Python integration tests and Rust integration tests: should I be adding both, or just python integration tests? 
3. Any good articles/documentation that you know of for signing commits with an already provided public GPG key? Everything online I seem to find about signing commits involves generating a key pair but that doesn't seem right to me since this repo already provides a public GPG key. I added the public GPG key in this repo to my GitHub account but I'm having a difficult time telling Git locally to use the GPG key to sign.

alindima commented:
> Oh no. I think I might've messed things up pretty badly while trying to get this develop branch in my fork rebased onto the latest main. I checked out main in my fork, did a pull from the upstream main branch, then checked out my local feature branch, and ran `git rebase main`. There were a few small conflicts during the rebase that I resolved. After that I tried pushing those changes to the feature branch of my fork but that had diverged with my local feature branch so I had to do a pull. After doing that and addressing the merge conflicts, things look pretty messed up. I'm still trying to figure out what happened and how to correct it so please bear with me until I get that figured out.

no worries. you can also create a new branch from latest main and cherry-pick the commits you've added on this branch.
You can then `reset --hard` this snapshot_vm_config branch to be identical to this new branch and force-push it.

alindima commented:
> A few questions:
> 
> 1. I noticed a coverage test failing and I feel like I recall reading somewhere in the documentation a while back that mentioned a need to update some coverage numbers in order to get some tests to pass. I can't seem to find any detailed information in the docs about updating the coverage numbers. Any idea there is some documentation describing what coverage numbers need to be updated and how?

Here is an example commit that updates the numbers: https://github.com/firecracker-microvm/firecracker/pull/2835/commits/cf72ba578da426acad5a3a4faa58c27a8271de6d

You need to update them if you see the coverage test reported as failed in the CI. The test will tell you what the current coverage is.
Note that the coverage dictionary is per-CPU type (ARM, Intel, AMD) and also kernel-version dependent. (So you may need to update a total of 6 values).

If you need help with this, I can show you how to update them once all other things are done and the PR is ready for merge.

As a side note, you need one of the maintainers to unblock the CI run. Feel free to ping me on slack if you need to have the CI unblocked and we don't notice it.

> 2. For integration tests, I've seen Python integration tests and Rust integration tests: should I be adding both, or just python integration tests?

In this specific case, python integration tests are what we're after. We want to test the API end-to-end. The python integration tests basically compile and run firecracker as a regular user would do.
Rust integration tests on the other hand test the Rust libraries/crates and cannot test the binary itself.

> 3. Any good articles/documentation that you know of for signing commits with an already provided public GPG key? Everything online I seem to find about signing commits involves generating a key pair but that doesn't seem right to me since this repo already provides a public GPG key. I added the public GPG key in this repo to my GitHub account but I'm having a difficult time telling Git locally to use the GPG key to sign.

Don't bother with GPG-signing your commits. Running `git commit -s` will add a `Signed-off-by` line to each commit, which is what our tests verify.

summerlinryan commented:
I'm working on getting this passing for existing tests and adding new tests but I have a question. How do you all debug the integration tests? I attempted to run `pytest --pdb` to get debugging but that didn't work because it required sudo.  Running `sudo pytest` doesn't work though because `pytest` isn't recognized as a command for sudo. I tried using `su` to temporarily get me in a root environment but then I no longer have access to the repo since the repo is stored in a folder in my users home directory. I must be overlooking something obvious here.

alindima commented:
> I'm working on getting this passing for existing tests and adding new tests but I have a question. How do you all debug the integration tests? I attempted to run `pytest --pdb` to get debugging but that didn't work because it required sudo. Running `sudo pytest` doesn't work though because `pytest` isn't recognized as a command for sudo. I tried using `su` to temporarily get me in a root environment but then I no longer have access to the repo since the repo is stored in a folder in my users home directory. I must be overlooking something obvious here.

We don't have a go-to way for debugging integration tests. Usually the output of the failing test is enough. You may output additional things using `print()` inside the test and capture the output when running with `pytest -s`.

Now I see that the tests that are failing on this PR are build and style tests - mostly unit tests.
In this case the output is pretty verbose and contains a lot of information.

Note that some performance tests will likely fail on your machine, because they assume more powerful machines (like AWS m5d.metal), but this is no issue as long as they pass in the CI.

We recommend running pytest inside devtool. You can get a root shell inside the devtool container like so:
`tools/devtool shell -p`

summerlinryan commented:
So I noticed that I accidentally violated some of the contribution quality standards by not ensuring all commits were formatted & passing unit tests. I've corrected that now in my most recent commits but I'm wondering if I should do a interactive rebase to ensure that each and every commit is formatted & passing unit tests. Also, I'm thinking I'll need to interactive rebase & force push to adjust my commit messages because of them are slightly exceeding the title length requirement of gitlint and missing signoff messages. Do both of those adjustments sound correct to you? 

Lastly, this is a small question, do you think the testing of the config when restoring a snapshot should go in the test_api.py test file or the test_snapshot_[basic/advanced].py files? I'm leaning towards the test_api.py but I could see it either way.

alindima commented on src/vmm/src/device_manager/persist.rs:
```diff
@@ -616,6 +650,8 @@ mod tests {
             vm: vmm.vm.fd(),
             event_manager: &mut event_manager,
+            for_each_restored_device: VmResources::update_from_restored_device,
+            vm_resources: &mut Default::default(),
```
In this test we can also assert the expected `vm_resources` configuration

alindima reviewed (commented):
We should also add a changelog entry, unit tests for the add_device/set_device methods and python integration tests.

Note that we have recently changed the MMDS API so it's best to rebase your branch again. In a nutshell, the net device no longer has the allow_mmds_requests.
It's best to do this periodically and to fix eventual conflicts.

alindima commented:
> So I noticed that I accidentally violated some of the contribution quality standards by not ensuring all commits were formatted & passing unit tests. I've corrected that now in my most recent commits but I'm wondering if I should do a interactive rebase to ensure that each and every commit is formatted & passing unit tests. Also, I'm thinking I'll need to interactive rebase & force push to adjust my commit messages because of them are slightly exceeding the title length requirement of gitlint and missing signoff messages. Do both of those adjustments sound correct to you?

Yes, you'll need to do an interactive rebase and modify the commit messages.
This PR has a lot of commits and most of them should be squashed.
We usually make sure that every commit successfully builds on its own and we aim for commits that each serve a defined purpose (for example, one commit for doc updates, one commit for tests, one commit for a specific unitary code change).
Commits that modify a piece of code introduced by the same PR should usually be squashed.

> Lastly, this is a small question, do you think the testing of the config when restoring a snapshot should go in the test_api.py test file or the test_snapshot_[basic/advanced].py files? I'm leaning towards the test_api.py but I could see it either way.

Let's have it in `test_api.py`, since that's where `test_get_full_config` is. You could actually edit that test to perform the validation both pre and post snapshot (to reuse most of the code)

summerlinryan commented:
@alindima I've addressed all of the recent feedback regarding the commits, unit & integration tests, and CHANGELOG entry. Let me know how it all looks when you get the chance, thanks!

summerlinryan commented on src/vmm/src/device_manager/persist.rs:
```diff
@@ -645,16 +646,74 @@ mod tests {
+    {{
+      "iface_id": "netif",
+      "host_dev_name": "hostname",
+      "guest_mac": "00:00:00:00:00:00",
```
@alindima  I'm unsure what to think about this guest_mac being all 0s. It feels like it should actually be null because the guest_mac is None on line 604 but then I saw some code in Net::new_with_tap that seemed to use a ConfigSpace::default() which has all 0s for the mac address so then I figured maybe the all 0s is right.

alindima replied:
Nice catch! I've done some digging to fond out why that is.
This is all zeros because indeed the guest mac in the config space is initialised with zeros. When the Net device is then saved to snapshot, we don't specifically save the guest mac as configured by the user and we then initialise it from the config space value, which in this case is all zeros.

I think this is a bug so we should open a new issue for it after we merge the PR. It is confusing to have a zeroed MAC when it was not really configured.

summerlinryan replied:
@alindima You want to wait until after this PR is merged before we create the new issue related to the zeroed out mac address, right? Just going through and making sure I've addressed all outstanding comments.

alindima replied:
Let's wait until we merge this PR and leave a comment on the same issue with the unaddressed items, like this MAC issue and the fact that the cpu_template, boot source and smt are currently not persisted.

alindima commented:
> @alindima I've addressed all of the recent feedback regarding the commits, unit & integration tests, and CHANGELOG entry. Let me know how it all looks when you get the chance, thanks!

great! I'll have a look probably by early next week

summerlinryan commented:
Although I thought I verified that all tests passed locally before pushing, I must've made a mistake because I see that the python/markdown linting are failing. I'll fix those small issues when I get home from work tonight.

alindima commented on src/vmm/src/device_manager/persist.rs:
```diff
@@ -611,14 +646,74 @@ mod tests {
 
+        let expected_vm_resources = format!(
+            r#"{{
+  "balloon": {{
```
the indentation is a bit off

summerlinryan replied:
If I change the indentation here, the test will fail since it's a raw string and the indentation of the raw string JSON will no longer match the pretty printed JSON from serde serialization. Is there another way to change the indentation formatting but still have the assertion pass?

summerlinryan commented on src/vmm/src/resources.rs:
```diff
@@ -72,7 +73,7 @@ pub struct VmmConfig {
     block_devices: Vec<BlockDeviceConfig>,
     #[serde(rename = "boot-source")]
-    boot_source: BootSourceConfig,
+    boot_source: Option<BootSourceConfig>,
```
I'm having a hard time finding it now but there was some documentation edit suggestion in the past that used the word "empty" when referring to a couple of properties (one of which was `boot_source`). That made me think that in the case of restored snapshots, we'd want the `boot_source` to appear as null (since there isn't actually a boot source in that case). If we'd rather keep it as the default `BootSourceConfig` that's no problem, just let me know and I'll undo this change of making it `Optional`.

summerlinryan commented on src/vmm/src/resources.rs:
```diff
@@ -94,7 +95,7 @@ pub struct VmResources {
     vm_config: VmConfig,
     /// The boot configuration for this microVM.
-    boot_config: Option<BootConfig>,
+    pub boot_config: Option<BootConfig>,
```
No, it actually doesn't. This was from a previous failed approach of fixing this bug and accidentally got left in. Sorry for leaving it in and nice job spotting it.

summerlinryan commented on tests/integration_tests/functional/test_api.py:
```diff
@@ -1407,175 +1407,18 @@ def test_get_full_config(test_microvm_with_api):
+        mem_file_path='memfile',
+        snapshot_path='statefile',
+        diff=False,
+        version='0.24.0'
```
Honestly, I can't recall why I explicitly set the version to 0.24.0. I must've just been confused. I'll update it to to be unspecified (using latest).

alindima replied:
sounds good, let's remove it

summerlinryan commented on src/vmm/src/vmm_config/vsock.rs:
```diff
@@ -192,4 +192,20 @@ pub(crate) mod tests {
+        .unwrap();
+
+        vsock_builder.set_device(Arc::new(Mutex::new(vsock)));
+        assert!(vsock_builder.inner.is_some());
```
I added an assertion for the `uds_path` of the `vsock_builder`. I have assertions for the ids of all of the `test_add_device` tests that involve builders with collections. Is there something else I can assert to confirm the identity of the device for the builders with collections (e.g. `BlockBuilder`)? The only other test that I can think of that you might be referring to would be the `BalloonBuilder.test_set_device` but I'm not sure what I can check in terms of identity on that because the `Balloon.id()` always returns `BALLOON_DEV_ID` so checking the `id` of the device wouldn't be very useful in that case.

summerlinryan commented:
I made most of the recent feedback changes locally and have rebased on to main so I went ahead and marked some of your feedback comments as resolved. I'm going to wait to push the changes since I might be making a change to the Optional `boot_source` depending on your feedback to my other comment.

alindima commented:
> I made most of the recent feedback changes locally and have rebased on to main so I went ahead and marked some of your feedback comments as resolved. I'm going to wait to push the changes since I might be making a change to the Optional `boot_source` depending on your feedback to my other comment.

Great, thanks! I've responded to the comments.
One other outstanding item is the integration test. View this comment: https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r799571125

summerlinryan commented on tests/integration_tests/functional/test_api.py:
```diff
@@ -1342,6 +1342,142 @@ def test_api_balloon(test_microvm_with_api):
+    snapshot_builder = SnapshotBuilder(test_microvm)
+
+    # Create base snapshot.
+    snapshot = snapshot_builder.create([root_disk.local_path()],
```
@alindima I'm a bit confused on this final test. I've been trying to find tests to reference for creating a microvm from a snapshot (e.g. the tests in `test_snapshot_basic.py`) and they mostly seem to use context to get the `ssh_key`, `root_disk`, and `snapshot_type` but when I try to pass `context` into the new test I made, I get an error that I can't get past. This approach that I'm trying is using the `MicrovmBuilder` and `SnapshotBuilder` but I also tried another approach where I was trying to use `SnapshotHelper.load` but that failed as well. Any tips on which approach would be the best? Thanks!

alindima replied:
This is a little tricky to get right. Try applying this patch:
```
diff --git a/tests/integration_tests/functional/test_api.py b/tests/integration_tests/functional/test_api.py
index be039c84..a0b66a84 100644
--- a/tests/integration_tests/functional/test_api.py
+++ b/tests/integration_tests/functional/test_api.py
@@ -19,7 +19,7 @@ import host_tools.network as net_tools
 from conftest import _test_images_s3_bucket
 
 from framework.utils import is_io_uring_supported
-from framework.artifacts import ArtifactCollection, SnapshotType
+from framework.artifacts import ArtifactCollection, SnapshotType, NetIfaceConfig, DEFAULT_DEV_NAME, DEFAULT_TAP_NAME
 from framework.builder import MicrovmBuilder, SnapshotBuilder
 
 MEM_LIMIT = 1000000000
@@ -1342,19 +1342,24 @@ def test_api_balloon(test_microvm_with_api):
     assert test_microvm.api_session.is_status_bad_request(response.status_code)
 
 
-def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_cloner_path):
+def test_get_full_config_after_restoring_snapshot(bin_cloner_path):
     """
     Test the reported configuration of a microVM configured with all resources.
 
     @type: functional
     """
-    test_microvm = test_microvm_with_api
+    microvm_builder = MicrovmBuilder(bin_cloner_path)
+    test_microvm = microvm_builder
+
+    net_iface = NetIfaceConfig()
+    vm_instance = microvm_builder.build_vm_nano(net_ifaces=[net_iface])
+    test_microvm = vm_instance.vm
+    root_disk = vm_instance.disks[0]
+    ssh_key = vm_instance.ssh_key
 
     setup_cfg = {}
 
-    test_microvm.spawn()
     # Basic config also implies a root block device.
-    test_microvm.basic_config(rootfs_io_engine="Sync")
     setup_cfg['machine-config'] = {
         'vcpu_count': 2,
         'mem_size_mib': 256,
@@ -1362,13 +1367,10 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         'track_dirty_pages': False,
         'cpu_template': 'C3',
     }
-    setup_cfg['boot-source'] = {
-        'kernel_image_path': '/vmlinux.bin',
-        'initrd_path': None
-    }
+    test_microvm.machine_cfg.patch(**setup_cfg['machine-config'])
     setup_cfg['drives'] = [{
         'drive_id': 'rootfs',
-        'path_on_host': '/bionic.rootfs.ext4',
+        'path_on_host': f"/{os.path.basename(root_disk.local_path())}",
         'is_root_device': True,
         'partuuid': None,
         'is_read_only': False,
@@ -1397,11 +1399,20 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         'uds_path': 'vsock.sock'
     }
 
-    # Add a net device.
-    iface_id = '1'
-    tapname = test_microvm.id[:8] + 'tap' + iface_id
-    tap1 = net_tools.Tap(tapname, test_microvm.jailer.netns)
-    guest_mac = '06:00:00:00:00:01'
+    setup_cfg['logger'] = None
+    setup_cfg['metrics'] = None
+    setup_cfg['mmds-config'] = {
+        'version': "V1",
+        'network_interfaces': [DEFAULT_DEV_NAME]
+    }
+
+    response = test_microvm.mmds.put_config(json=setup_cfg['mmds-config'])
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    # Start the microvm.
+    test_microvm.start()
+
+    # Add a tx rate limiter to the net device.
     tx_rl = {
         'bandwidth': {
             'size': 1000000,
@@ -1410,46 +1421,19 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         },
         'ops': None
     }
-    response = test_microvm.network.put(
-        iface_id=iface_id,
-        guest_mac=guest_mac,
-        host_dev_name=tap1.name,
+    response = test_microvm.network.patch(
+        iface_id=DEFAULT_DEV_NAME,
         tx_rate_limiter=tx_rl
     )
     assert test_microvm.api_session.is_status_no_content(response.status_code)
     setup_cfg['network-interfaces'] = [{
-        'iface_id': iface_id,
-        'host_dev_name': tap1.name,
-        'guest_mac': '06:00:00:00:00:01',
+        'guest_mac': net_tools.mac_from_ip(net_iface.guest_ip),
+        'iface_id': DEFAULT_DEV_NAME,
+        'host_dev_name': DEFAULT_TAP_NAME,
         'rx_rate_limiter': None,
         'tx_rate_limiter': tx_rl
     }]
 
-    # Update MMDS config.
-    mmds_config = {
-        'version': 'V2',
-        'ipv4_address': '169.254.169.250',
-        'network_interfaces': ['1']
-    }
-    response = test_microvm.mmds.put_config(json=mmds_config)
-    assert test_microvm.api_session.is_status_no_content(response.status_code)
-
-    setup_cfg['logger'] = None
-    setup_cfg['metrics'] = None
-    setup_cfg['mmds-config'] = {
-        'version': 'V2',
-        'ipv4_address': '169.254.169.250',
-        'network_interfaces': ['1']
-    }
-
-    # Start the microvm.
-    test_microvm.start()
-
-    microvm_builder = MicrovmBuilder(bin_cloner_path)
-
-    root_disk = test_microvm.vm.disks[0]
-    ssh_key = test_microvm.ssh_key
-
     # Create a snapshot builder from a microvm.
     snapshot_builder = SnapshotBuilder(test_microvm)
 
@@ -1465,11 +1449,12 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
     # We expect boot-source, machine-config.smt, and machine-config.cpu_template to all
     # be empty/default after restoring from a snapshot.
     expected_cfg['boot-source'] = {
-        'kernel_image_path': None,
+        'kernel_image_path': '',
         'initrd_path': None
     }
     expected_cfg['machine-config']['smt'] = False
-    expected_cfg['machine-config']['cpu_template'] = None
+    expected_cfg['machine-config'].pop('cpu_template')
+    expected_cfg['mmds-config'] = None
 
     # Validate full vm configuration post-restore.
     response = microvm.full_cfg.get()
```

alindima replied:
Notice that the mmds_config is still reported as empty after snapshotting, you need to also hook that in when re-adding the network device to VmResources.
There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

summerlinryan replied:
Unfortunately, some stuff came up in my personal life. I'll look back into this soon (should be within a few days).

alindima replied:
Sure, no worries!

summerlinryan replied:
Sorry I'm struggling so much on this final test. I'm not sure what you mean with the following:

> you need to also hook that in when re-adding the network device to VmResources.

> There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

It seems that the MMDS config always shows as null (which is maybe what you're saying)? If it should be null after snapshot, then I'm thinking what I have currently, is right.

Also, any idea what I can use to test the `cpu_template` of the `machine-config`? I had it set to `C3` because that's what some other tests were using but I was getting an `InvalidVendor` that was caused by the C3 `cpu_template` so I ended up getting rid of setting the `cpu_template` (at least temporarily). 

As a side not, I'll merge the 2 final tests commits together this is worked through. I just wanted to push what I had for review.

alindima replied:
> Sorry I'm struggling so much on this final test.

No worries. Don't get discouraged, you've done the more difficult part 😄 .

> It seems that the MMDS config always shows as null (which is maybe what you're saying)?

Yes, that what I was referring to. It should not be empty after restoring from snapshot. The `NetState` struct contains the optional `MmdsNetworkStackState` which can be used to extract the Mmds config. This should be done in the `update_from_restored_device` function, for each of the net devices.

> There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

What I meant by this is that the used MMDS version is not yet saved in the snapshot (so it should always be V1, the default).

alindima replied:
> Also, any idea what I can use to test the cpu_template of the machine-config? I had it set to C3 because that's what some other tests were using but I was getting an InvalidVendor that was caused by the C3 cpu_template so I ended up getting rid of setting the cpu_template (at least temporarily).

Hmm, are you using an AMD host by any chance?

alindima replied:
You should use the `utils.get_cpu_vendor()` in the test, and only set that CPU template for intel hosts

summerlinryan replied:
> Hmm, are you using an AMD host by any chance?

Yes I am actually. I'll use your recommendation of using `utils.get_cpu_vendor` to conditionally set the `cpu_template`. Thanks!

After rereading a bunch of MMDS documentation, reading through the code for a while, and experimenting with the API, I feel like I'm understanding the MMDS change more than I was before. A few implementation questions I still have though are:

1. You mentioned using `MmdsNetworkStackState` in the `NetState`. I noticed that property is currently not public; are we comfortable making it public? On a related note though, the `update_from_restored_device` currently doesn't have access to the `NetState` but it does have access to the `Net` device which seems to have the MMDS network stack stored on it so I'm thinking about using that instead of the `NetState`. Does that sound reasonable or is there a reason I should still make use `NetState`? If it sounds reasonable to use `Net` instead of `NetState`, then ignore my first question about making the `NetState` MMDS state public. 
2. There is only one `ipv4_address` that exists for the `MmdsConfig`, however, each `Net` device has a `ipv4_addr` that seems to refer to the MMDS server. Presumably, the `ipv4_addr` for all net devices will always be the same (since there is only a single MMDS server) but correct me if that presumption is wrong. What I'm thinking about doing is setting the `MmdsConfig` ipv4 address to the default (`169.254.169.254`) in `build_microvm_from_snapshot` and then updating it to the value of the `ipv4_addr` of the net device in `update_from_restored_device`. Does that sound reasonable? (see the code snippet below for an example of how I might go about the update in `update_from_restored_device`)

```rs
 SharedDeviceType::SharedNetwork(network) => {
    if let Some(mmds_ns) = &network.lock().unwrap().mmds_ns {
        let mmds_config = self.mmds_config.as_mut().unwrap();
        mmds_config.ipv4_address = Some(mmds_ns.ipv4_addr);
        mmds_config
            .network_interfaces
            .push(network.lock().expect("Poisoned lock").id().to_owned());
    }

    self.net_builder.add_device(network);
}
```

alindima replied:
> You mentioned using MmdsNetworkStackState in the NetState. I noticed that property is currently not public; are we comfortable making it public? On a related note though, the update_from_restored_device currently doesn't have access to the NetState but it does have access to the Net device which seems to have the MMDS network stack stored on it so I'm thinking about using that instead of the NetState. Does that sound reasonable or is there a reason I should still make use NetState? If it sounds reasonable to use Net instead of NetState, then ignore my first question about making the NetState MMDS state public.

This was a mistake from me, I was indeed referring to the `Net` struct.

> There is only one ipv4_address that exists for the MmdsConfig, however, each Net device has a ipv4_addr that seems to refer to the MMDS server. Presumably, the ipv4_addr for all net devices will always be the same (since there is only a single MMDS server) but correct me if that presumption is wrong.

This is true 👍🏻 

> What I'm thinking about doing is setting the MmdsConfig ipv4 address to the default (169.254.169.254) in build_microvm_from_snapshot and then updating it to the value of the ipv4_addr of the net device in update_from_restored_device. Does that sound reasonable? (see the code snippet below for an example of how I might go about the update in update_from_restored_device)

No need to set the ipv4 address to default before `update_from_restored_device` (since the `ipv4_addr` field on the `MmdsNetworkStack` is not optional; if there is no network interface with MMDS configured, then there is no MMDS configuration at all). The code snippet you pasted looks good to me overall.

Note that `self.mmds_config.as_mut().unwrap()` may panic as it's written in the snippet.

summerlinryan commented:
@alindima I think I've got the MMDS config working as expected now and I've updated the tests. I've also updated test to conditionally set the cpu template only for Intel hosts. Unfortunately, there were 2 fields I had to change from `pub(crate)` to `pub` but it seemed necessary in order to restore the MMDS config. Let me know what you think about everything, thanks!

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -171,6 +172,44 @@ impl VmResources {
+                if let Ok(net) = network.lock() {
+                    if let Some(mmds_ns) = &net.mmds_ns {
+                        if let Some(mut_mmds_config) = self.mmds_config.as_mut() {
+                            mut_mmds_config.ipv4_address = Some(mmds_ns.ipv4_addr);
```
we may skip this since they should all have the same ipv4 adddress

alindima commented on src/vmm/src/resources.rs:
```diff
@@ -171,6 +172,44 @@ impl VmResources {
+            }
+
+            SharedDeviceType::SharedNetwork(network) => {
+                if let Ok(net) = network.lock() {
```
let's do an `expect("Poisoned lock")` here, since this is an irrecoverable error

summerlinryan replied:
I changed it to use the `expect("Poisoned lock")` but doing that introduced a borrow checker issue since `network.lock().expect()` results in a borrow but then `network` gets moved with `self.net_builder.add_device(network);`. A workaround I added was a putting the `if` in its own scope but I feel like there must be a better , more idiomatic Rust way? I tried cloning the `network` device but that also had a separate borrow checker issue.

alindima replied:
you may call `drop(net)` before calling `self.net_builder.add_device(network);`, but I think the current approach is fine

alindima commented on CHANGELOG.md:
```diff
@@ -99,6 +99,10 @@
 
+- GET `/vm/config` was returning a default config object after restoring from a snapshot.
+  It now correctly returns the config of the original microVM, except for boot_config and the
+  cpu_template and smt fields of the machine config,
+  which are currently lost.
```
```suggestion
  cpu_template and smt fields of the machine config, which are currently lost.
```

alindima commented on src/api_server/swagger/firecracker.yaml:
```diff
@@ -612,7 +612,8 @@ paths:
       description:
-        Gets configuration for all VM resources.
+        Gets configuration for all VM resources. If the VM is restored from a snapshot, the boot-source,
+        machine-config.ht_enabled and machine-config.cpu_template will be empty.
```
```suggestion
        machine-config.smt and machine-config.cpu_template will be empty.
```

alindima commented:
> @alindima I think I've got the MMDS config working as expected now and I've updated the tests. I've also updated test to conditionally set the cpu template only for Intel hosts. Unfortunately, there were 2 fields I had to change from `pub(crate)` to `pub` but it seemed necessary in order to restore the MMDS config. Let me know what you think about everything, thanks!

Looks good @summerlinryan . Please have a look at the failing CI tests and try to fix them. Let me know if you need help

dianpopa commented:
I think we should do a full fix for this issue and implement the right solution which would be to save all the info related to the vm config in a separate structure and store that in the snapshot. We would need to be careful about not breaking snapshot compatibility by checking whether the vm config info is available in the snapshot. Sorry if this was discussed eariler, there are lots of comments and I might have missed some of them. 

WDYT about the solution proposed?

alindima commented:
> I think we should do a full fix for this issue and implement the right solution which would be to save all the info related to the vm config in a separate structure and store that in the snapshot. We would need to be careful about not breaking snapshot compatibility by checking whether the vm config info is available in the snapshot. Sorry if this was discussed eariler, there are lots of comments and I might have missed some of them. 
> 
> WDYT about the solution proposed?

We would only need to do such things for smt and cpu template (and potentially boot source, but I don't really think it adds any value)
We decided to move on with this PR and skip those for now.

We can definitely add those as a followup

dianpopa commented:
> > I think we should do a full fix for this issue and implement the right solution which would be to save all the info related to the vm config in a separate structure and store that in the snapshot. We would need to be careful about not breaking snapshot compatibility by checking whether the vm config info is available in the snapshot. Sorry if this was discussed eariler, there are lots of comments and I might have missed some of them.
> > WDYT about the solution proposed?
> 
> We would only need to do such things for smt and cpu template (and potentially boot source, but I don't really think it adds any value) We decided to move on with this PR and skip those for now.
> 
> We can definitely add those as a followup

It would be great to add them as a follow up but is there a way to obtain those info  after resuming from snapshot (smt maybe, but for cpu template I do not see a way...)? This is my main concern actually, that we would be left with a partial solution. 
Sorry for missing the decision discussion.

alindima commented:
> > > I think we should do a full fix for this issue and implement the right solution which would be to save all the info related to the vm config in a separate structure and store that in the snapshot. We would need to be careful about not breaking snapshot compatibility by checking whether the vm config info is available in the snapshot. Sorry if this was discussed eariler, there are lots of comments and I might have missed some of them.
> > > WDYT about the solution proposed?
> > 
> > 
> > We would only need to do such things for smt and cpu template (and potentially boot source, but I don't really think it adds any value) We decided to move on with this PR and skip those for now.
> > We can definitely add those as a followup
> 
> It would be great to add them as a follow up but is there a way to obtain those info after resuming from snapshot (smt maybe, but for cpu template I do not see a way...)? This is my main concern actually, that we would be left with a partial solution. Sorry for missing the decision discussion.

Here is where we discussed this topic: https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r769688210

In the current form of this PR, this would indeed be a partial solution (which fixes more than 90% let's say of the issue).
The followup would take the smt information and the cpu template likely from the snapshot state (which currently does not persist them).
Do you have a concern with merging such a partial solution and then coming back to it as a follow-up? Just as a clarification, the follow-up would build on the solution from this PR (the solutions are conflicting here AFAICT).

I believe that a partial solution is better than trying to fix everything in one go (given that it's a pretty complicated one as well)

dianpopa commented:
> > > > I think we should do a full fix for this issue and implement the right solution which would be to save all the info related to the vm config in a separate structure and store that in the snapshot. We would need to be careful about not breaking snapshot compatibility by checking whether the vm config info is available in the snapshot. Sorry if this was discussed eariler, there are lots of comments and I might have missed some of them.
> > > > WDYT about the solution proposed?
> > > 
> > > 
> > > We would only need to do such things for smt and cpu template (and potentially boot source, but I don't really think it adds any value) We decided to move on with this PR and skip those for now.
> > > We can definitely add those as a followup
> > 
> > 
> > It would be great to add them as a follow up but is there a way to obtain those info after resuming from snapshot (smt maybe, but for cpu template I do not see a way...)? This is my main concern actually, that we would be left with a partial solution. Sorry for missing the decision discussion.
> 
> Here is where we discussed this topic: [#2806 (comment)](https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r769688210)
> 
> In the current form of this PR, this would indeed be a partial solution (which fixes more than 90% let's say of the issue). The followup would take the smt information and the cpu template likely from the snapshot state (which currently does not persist them). Do you have a concern with merging such a partial solution and then coming back to it as a follow-up? Just as a clarification, the follow-up would build on the solution from this PR (the solutions are not conflicting here AFAICT).
> 
> I believe that a partial solution is better than trying to fix everything in one go (given that it's a pretty complicated one as well)

Alin and me synced offline and decided to go ahead with the solution proposed in this PR. @summerlinryan let us know if you need help with the feedback.

summerlinryan commented:
@dianpopa When you say "the solution proposed in this PR", are you referring to the work that's already been done or the new approach you suggested including the saving of the full config? If you're referring to the work that's already been done, then I should be working on getting the tests passing tonight so within the next day or so I should have the final commit pushed. 👍 Thanks for the feedback!

dianpopa commented:
> @dianpopa When you say "the solution proposed in this PR", are you referring to the work that's already been done or the new approach you suggested including the saving of the full config? If you're referring to the work that's already been done, then I should be working on getting the tests passing tonight so within the next day or so I should have the final commit pushed. 👍 Thanks for the feedback!

Oh sorry for the misunderstanding. I meant that we should go ahead with the work that's already been done.

summerlinryan commented:
> Please rebase over latest main

Sorry, I thought I had. I must've either forgot to or made a mistake while doing the rebase.

Nice catch on `setup_cfg['machine-config']['cpu_template']: 'C3'`! That was definitely a problem that I was completely overlooking. 

I updated the coverage and got all other tests passing as far I can tell locally. It seems that the `test_python` test doesn't work for me locally. I end up getting a ton of 

```
Exception ignored when trying to write to the signal wakeup fd:
BlockingIOError: [Errno 11] Resource temporarily unavailable
```

errors when running it locally and it seems to timeout. I'm assuming that's happening from how many processes it's spawning. 

Also, I wasn't sure how to test the `aarm64` `platform.machine()` change locally since I'm running on a `x86_64` host but it seemed like a simple change so hopefully I'll wake up in the morning to a bunch of passing tests! :crossed_fingers:

alindima commented:
Most of the tests are passing now, great! 👍🏻 

I see that on ARM there's a test that fails: `test_api_machine_config`.

This is because this test was relying on the logic that on ARM `get_cpu_vendor()` used to return Intel (this logic was wrong).

AFAICT, this change would fix things (in test_api.py):

`if utils.get_cpu_vendor() != utils.CpuVendor.INTEL:` => `if utils.get_cpu_vendor() == utils.CpuVendor.AMD:`

I believe you can also mark this PR as ready for review now (it's still a draft)

alindima commented:
Tests are all green! 😄 
@summerlinryan there are a couple of comments from me left unaddressed. @dianpopa will also review the code.
Looking good!

alindima reviewed (approved):
Looks great to me, thanks for the hard work! 😄

georgepisaltu commented:
Awesome work @summerlinryan 🎉
